pkg/exchange/binance/stream.go

```diff
@@ -146,34 +159,57 @@ func (s *Stream) handleDisconnect() {
 }
 
 func (s *Stream) handleConnect() {
+	var err error
 	if !s.PublicOnly {
-		// Emit Auth before establishing the connection to prevent the caller from missing the Update data after
-		// creating the order.
-
-		// spawn a goroutine to emit auth event to prevent blocking the main event loop
-		go s.EmitAuth()
-		return
+		if len(s.privateKey) == 0 {
+			log.Warnf("Ed25519 private key is not set, using deprecated HMAC authentication")
+			err = s.authConnectHMAC()
+		} else {
+			err = s.authConnectEd25519()
+		}
+	}
+	err = s.subscribe()
+	if err != nil {
+		log.WithError(err).Error("subscribe error")
 	}
+}
+
+func (s *Stream) authConnectHMAC() error {
+	go s.EmitAuth()
+	return nil
+}
 
+func (s *Stream) authConnectEd25519() error {
+	signature, timestamp := generateSignature(s.client.APIKey, s.privateKey)
```

Comment:
you might need to move this function to the binanceapi package later for reuse

Reply:
I already move the `generateSignature` to `pkg/exchange/binance/util.go`.
The RESTful client should have access to it.

Reply:
Oh, the `binanceapi` package.
I see.
Will do.

Reply:
Done.